src/mlpack/methods/ann/layer/multihead_attention.hpp

```diff
@@ -0,0 +1,394 @@
+ * parameters.
+ */
+class EncoderDecoderAttention
+{
+ public:
+
+  //! True if query, key and value are equal.
+  static const bool allInputsEqual = false;
+
+  //! True if key and value are equal.
+  static const bool keyEqualsValue = true;
+
+  /**
+   * Checks the validity of the dimensions of the input.
+   */
+  template <typename InputType>
+  static void CheckInputDimensions(const InputType& input,
+                           const size_t tgtSeqLen,
+                           const size_t srcSeqLen,
+                           const size_t embedDim)
+  {
+    Log::Assert(input.n_rows == (tgtSeqLen + srcSeqLen) * embedDim);
+  }
+
+  template <typename eT>
+  static void SplitInput(const arma::Mat<eT>& input,
+                         arma::Cube<eT>& query,
+                         arma::Cube<eT>& key,
+                         arma::Cube<eT>& value,
+                         const size_t tgtSeqLen,
+                         const size_t srcSeqLen,
+                         const size_t embedDim)
+  {
+    CheckInputDimensions(input, tgtSeqLen, srcSeqLen, embedDim);
+    query = arma::Cube<eT>(const_cast<arma::Mat<eT>&>(input).memptr(),
+        embedDim, tgtSeqLen, input.n_cols, false);
+    key = arma::Cube<eT>(const_cast<arma::Mat<eT>&>(input).memptr()
+        + query.n_elem, embedDim, srcSeqLen, input.n_cols, false);
+    value = key;
+  }
+};
+
+/**
+ * GeneralAttention policy class. It deals with the constraints of the
+ * parameters.
+ */
+class GeneralAttention
+{
+ public:
+
+  //! True if query, key and value are equal.
+  static const bool allInputsEqual = false;
+
+  //! True if key and value are equal.
+  static const bool keyEqualsValue = false;
+
+  /**
+   * Checks the validity of the dimensions of the input.
+   */
+  template <typename InputType>
+  static void CheckInputDimensions(const InputType& input,
+                                   const size_t tgtSeqLen,
+                                   const size_t srcSeqLen,
+                                   const size_t embedDim)
+  {
+    Log::Assert(input.n_rows == (tgtSeqLen + 2 * srcSeqLen) * embedDim);
+  }
+
+  template <typename eT>
+  static void SplitInput(const arma::Mat<eT>& input,
+                         arma::Cube<eT>& query,
+                         arma::Cube<eT>& key,
+                         arma::Cube<eT>& value,
+                         const size_t tgtSeqLen,
+                         const size_t srcSeqLen,
+                         const size_t embedDim)
+  {
+    CheckInputDimensions(input, tgtSeqLen, srcSeqLen, embedDim);
+    query = arma::Cube<eT>(const_cast<arma::Mat<eT>&>(input).memptr(),
+        embedDim, tgtSeqLen, input.n_cols, false);
+    key = arma::Cube<eT>(const_cast<arma::Mat<eT>&>(input).memptr()
+        + query.n_elem, embedDim, srcSeqLen, input.n_cols, false);
+    value = arma::Cube<eT>(const_cast<arma::Mat<eT>&>(input).memptr()
+        + query.n_elem + key.n_elem, embedDim, srcSeqLen, input.n_cols, false);
```

Comment:
I was wondering if it would be better to assume that each attention layer takes three inputs: query, key, and value.
In that case
1. You don't need to write different attention policies.
2. The code will be a bit more simple.
3. It would be easier to work with the layer due to the unique interface. Otherwise you have to write a method in order to pack the input for each attention type.

On the contrary
1. Passing three matrices to each attention layer could be a bit slower since you have to copy the query and the key and concatenate it to the input.

Reply:
>2. The code will be a bit more simple.

Yes, I totally agree. In fact, I started with that but due to the possibility of loss of efficiency I switched.
I will try that again.

Reply:
@lozhnikov: Done.